### ISSUE TITLE
Plugin compiled with Groovy's @CompileStatic breaks with Gradle 4.9

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#6027]
A bug in the Groovy compiler 2.4.10 and before ((GROOVY-8142)[https://issues.apache.org/jira/browse/GROOVY-8142]) can cause a plugin to refer to Gradle internal APIs even though its source only refers to public ones. Once Gradle changes or removes the unintentionally referred-to internal APIs, the published plugin stops working.

This is what happened at least in one case with the [Gradle Packer plugin](https://gitlab.com/ysb33rOrg/packer-gradle-plugin/blob/RELEASE_0_1_3/src/main/groovy/org/ysb33r/gradle/packer/tasks/PackerBuilder.groovy#L193): it refers to the public `Project.file()` API, but ended up with the bytecode referring to the internal `FileOperations.file()` API. In Gradle 4.9 `DefaultProject` stopped implementing `FileOperations`, which was a change to only internal APIs, yet it caused the Packer plugin to break with the following error:

```text
java.lang.IncompatibleClassChangeError: Class org.gradle.api.internal.project.DefaultProject_Decorated does not implement the requested interface org.gradle.api.internal.file.FileOperations
```

The problem to occur requires the following conditions to all be present:

- the plugin must be compiled with Gradle 3.5 or earlier (Gradle 4.0 shipped a year ago with Groovy 2.4.11 that has the compiler bug fixed)
- the plugin must be calling `Project.file()`, `Project.files()` etc. _from inside a type that extends `DefaultTask`_ (there might be other cases, but the methods on `FileOperations` are the ones that we know about)
- the task in question must be written in Groovy, and must be using `@CompileStatic`
- GROOVY-8142 must trigger while building the plugin – this doesn't happen every time, so even if all the above conditions are present, you might still end up with the correct bytecode

**Comment by lptr** [comment]
One way to fix a broken plugin is to re-build it using Gradle 4.0 or later. If the plugin must be compiled using an earlier Gradle version for backwards compatibility reason, the solution could be to recompile with a newer Groovy version.

Adding the following snippet should achieve exactly this:

```gradle
// Use a newer Groovy compiler to work around https://issues.apache.org/jira/browse/GROOVY-8142
// Only needed when running with Gradle versions earlier than 4.0
if (GradleVersion.current() >= GradleVersion.version("4.0")) {
    throw new RuntimeException("${GradleVersion.current()} already ships with a Groovy compiler that has GROOVY-8142 fixed")
} else {
    configurations {
        fixedGroovyCompiler
    }

    dependencies {
        fixedGroovyCompiler "org.codehaus.groovy:groovy-all:2.4.12"    
    }

    tasks.withType(GroovyCompile) {
        groovyClasspath = configurations.fixedGroovyCompiler
    }
}
```

**Comment by lptr** [comment]
Submitted a merge-request for the Packer plugin to fix it: https://gitlab.com/ysb33rOrg/packer-gradle-plugin/merge_requests/3

**Comment by lptr** [comment]
Closing this issue since there is a workaround to fix it for existing plugins.

**Comment by mark-vieira** [comment]
It seems this is still an issue. At least with Schalk's packer plugin.

https://gradle.com/s/y6jzkrnafxffw

**Comment by wolfs** [comment]
Seems like the asciidoctor plugin also ran into this problem: https://github.com/asciidoctor/asciidoctor-gradle-plugin/commit/7e0c9ffb8902cd4a5210fcc007d33f1afe63c63c
See https://github.com/asciidoctor/asciidoctor-gradle-plugin/issues/253#issuecomment-410207568

**Comment by lptr** [comment]
OK, so it seems Groovy 2.4.11 has fixed the reproducibility issue with which type a certain method is referred to when there are multiple candidates... But it now always uses that of the private field. This code for example:

```groovy
    /** The template file to use for the images.
     *
     * @return Resolved template file
     */
    @InputFile
    File getTemplate() {
        project.file(templateFile)
    }
```
https://gitlab.com/ysb33rOrg/packer-gradle-plugin/blob/master/src/main/groovy/org/ysb33r/gradle/packer/tasks/PackerBuilder.groovy#L154-161

Compiles into this:

```text
  public java.io.File getTemplate();
    descriptor: ()Ljava/io/File;
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=1, args_size=1
         0: aload_0
         1: checkcast     #2                  // class org/ysb33r/gradle/packer/tasks/PackerBuilder
         4: invokevirtual #42                 // Method getProject:()Lorg/gradle/api/Project;
         7: checkcast     #44                 // class org/gradle/api/internal/project/ProjectInternal
        10: aload_0
        11: getfield      #167                // Field templateFile:Ljava/lang/Object;
        14: invokeinterface #173,  2          // InterfaceMethod org/gradle/api/internal/file/FileOperations.file:(Ljava/lang/Object;)Ljava/io/File;
        19: areturn
        20: nop
        21: athrow
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      20     0  this   Lorg/ysb33r/gradle/packer/tasks/PackerBuilder;
      LineNumberTable:
        line 160: 0
      StackMapTable: number_of_entries = 1
        frame_type = 255 /* full_frame */
          offset_delta = 20
          locals = []
          stack = [ class java/lang/Throwable ]
    RuntimeVisibleAnnotations:
      0: #165()
```

Notice in particular:

```text
        14: invokeinterface #173,  2          // InterfaceMethod org/gradle/api/internal/file/FileOperations.file:
```

The same behavior is observed with Groovy 2.4.12 and 2.4.15.

**Comment by lptr** [comment]
Filed https://issues.apache.org/jira/browse/GROOVY-8732 for Groovy. It appears to be a problem on every release branch (2.4, 2.5, 2.6 and 3.0) alike.

To provide a workaround for affected plugins, we'll revert to implementing `FileOperations` in `DefaultProject` again in 4.10. Users who bump into the problem using Gradle 4.9 should upgrade to Gradle 4.10 as soon as it's out, or downgrade to Gradle 4.8.1.

We can introduce a better workaround in Gradle 5.0 if we stop using the same name for private fields and public getters where the private field has a more specific type. One example is the `AbstractTask.project` field.

**Comment by ysb33r** [comment]
Thanks @lptr for looking into this again - it had me scratching my head.

**Comment by big-guy** [comment]
Can we update TaskSubclassingBinaryCompatibilityCrossVersionSpec to cover this scenario?

**Comment by lptr** [comment]
@big-guy yes.

**Comment by lptr** [comment]
PR has been merged.

**Comment by lptr** [comment]
Looks like the original issue is going to be resolved in Groovy 2.5.2
